### PR TITLE
Add pyproject.toml, update installation instructions for modern Python packaging, fix docbuild, update CI scripts, add Cython 3 support

### LINKED ISSUE #93: Needs pyproject.toml to declare build-time dependency on cysignals
cypari2 is affected by https://github.com/pypa/pip/issues/6193

Even though dependencies are declared in `setup.py`, by the latest pip build isolation policies, it does not install `cysignals` before doing `bdist_wheel`.

### PR DESCRIPTION
**Pull request by mkoeppe**
This PR started out with minimal changes, replacing the outdated use of `setup_requires`.

Fixes #93 
(with a cleaner solution for augmenting the build-time path than the stalled PR #104).

### CODE REVIEW

File: _custom_build_meta.py
Comment (videlec) [inline _custom_build_meta.py:1]: Why do you make the file name starts with an underscore here?

Comment (mkoeppe) [reply _custom_build_meta.py:1]: No reason, feel free to rename

Comment (mkoeppe) [reply _custom_build_meta.py:1]: Do you want me to just remove the underscore, or do you have other naming preferences?

**Comment by dimpase** [comment]
@videlec - what is the status of cypari2 vs Cython 3? You probably noticed that Sage has moved to Cython 3 (and it works)

File: pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [build-system]
 requires = ["setuptools",
-            "Cython>=0.29",
+            "Cython>=0.29,<3",
```
Comment (dimpase) [inline pyproject.toml:3]: why? Cython 3 appears to work with cypari2.

Comment (videlec) [reply pyproject.toml:3]: It should. I was wondering were these weird compilation errors came from. I will remove the commit.

File: _custom_build_meta.py
Comment (videlec) [reply _custom_build_meta.py:1]: Nothing beyond following standard practice.

**Comment by dimpase** [comment]
@mkoeppe - can you add https://github.com/sagemath/sage/blob/develop/build/pkgs/cypari/patches/cython3-legacy.patch and drop the <3 constraint on Cython version added in [c473052](https://github.com/sagemath/cypari2/pull/139/commits/c47305286ece31526ba88fd65b0811a106856f4d) ?

File: pyproject.toml (hunk 1 above)
Comment (dimpase) [reply pyproject.toml]: Already done

**Comment by videlec** [comment]
I think the workflow should include both `Cython>=3` and `Cython<3`.

**Comment by dimpase** [comment]
> I think the workflow should include both `Cython>=3` and `Cython<3`.

This would be a bigger change than this PR. A new issue?

**Comment by dimpase** [comment]
@mkoeppe - what this `ERROR: Could not find a version that satisfies the requirement setuptools (from versions: none)` 
stuff mean? I have no clue.

**Comment by dimpase** [comment]
Does `make build` run an isolated Python environment, so that the previously run `pip install` has no effect?
I don't understand.

**Comment by dimpase** [comment]
@mkoeppe - it doesn't fly. `make build` does not want to accept system-wide installed `setuptools`. full stop.

**Comment by dimpase** [comment]
OK, we finally get a working build and testing, doc building is broken:
```
Running Sphinx v7.2.6
WARNING: Invalid configuration value found: 'language = None'. Update your configuration to a valid language code. Falling back to 'en' (English).
making output directory... done
WARNING: html_static_path entry '_static' does not exist
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 7 source files that are out of date
updating environment: [new config] 7 added, 0 changed, 0 removed

Exception occurred:
  File "/home/runner/work/cypari2/cypari2/tvenv/lib/python3.11/site-packages/sphinx/ext/extlinks.py", line 108, in role
    title = caption % part
            ~~~~~~~~^~~~~~
TypeError: not all arguments converted during string formatting
The full traceback has been saved in /tmp/sphinx-err-dpid6l08.log, if you want to report the issue to the developers.
```

**Comment by dimpase** [comment]
this is due to broken format strings in `extlinks` in `docs/source/conf.py`. fixing...

**Comment by dimpase** [comment]
OK, builds/ytests/docs with Pari 2.15.4 pass, ones with Pari 2.11 and 2.13 take much more time, no idea why.

**Comment by mkoeppe** [comment]
Let's get rid of `make build`. This makes no sense any more.

File: _custom_build_meta.py
Comment (mkoeppe) [reply _custom_build_meta.py:1]: You're making it sound as if you think my PR was not following standard practice.

**Comment by mkoeppe** [comment]
All done, please merge

**Pull request merged**